### ISSUE TITLE
V4.2.0 router events timing 

### ISSUE THREAD

**Issue by AmirSasson** [angular/angular#17473]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
since version 4.2.0 (4.2.0-rc.0 is broken, 4.2.0-beta.1 works) ngOnInit of Components are called only after router events are already been handled.
we have an application that  contains components that register to router on their ngOnInit(). till version 4.1.3 we used to get a the events of the router on ngOnInit() but it seems that this behavior was changed.
also it seems that App.Module components _do_ get the router events on ngOnInit(), but lower level modules don't. it is a great deal to us as the Router event is kind of an entry point for the entire module.

**Expected behavior**
same as v4.1.3 get notified for router events after ngOnInit() was called.

**Minimal reproduction of the problem with instructions**
https://plnkr.co/q9z9Rn 
see "Sub Module component router events" doesnt get router events at all

**What is the motivation / use case for changing the behavior?**
this is a breaking change 

**Please tell us about your environment:**
@angular/cli: 1.0.0
node: 8.0.0
os: win32 x64
@angular/animations: 4.2.2
@angular/common: 4.2.2
@angular/compiler: 4.2.2
@angular/core: 4.2.2
@angular/forms: 4.2.2
@angular/http: 4.2.2
@angular/platform-browser: 4.2.2
@angular/platform-browser-dynamic: 4.2.2
@angular/router: 4.2.2
@angular/cli: 4.2.2
@angular/compiler-cli: 4.2.2
@angular/language-service: 4.2.2

* **Browser:** [all]

 
* **Language:** [ TypeScript 2.3]

**Comment by fillmorejl** [comment]
I agree this is defiantly a breaking change when using ngOnInit on sub components.

In my case I send the child to the parent to render a header and actions buttons via an rxjs Subject in ngOnInit. This worked fine in 4.1.3, just updated to 4.2.2 and it is 100% broken requiring me to double router links for it to update the screen.

**Comment by kemsky** [comment]
It was indeed breaking change, but I'm not sure what has to be done, since it is already broken and fixed, reverting will cause another breaking change. 

Events are tricky, router should not change order or timing of events.

**Comment by fillmorejl** [comment]
I did not really "fix" is so to say, more of a workaround, reverting the change they made wont break my code, I will just move my code from the constructor back to ngOnInit when I have time, whether they change it back or not my code will remain operational.

Sadly when ngOnInit is fired on a sub component via triggering a sub route it seems to fire after the screen has been rendered which is causing many issues when adopting the use of ngOnInit.

**Comment by jasonaden** [comment]
The change was made to support animations. I understand how this could look like a breaking change, but there isn't a guarantee of the sequence of router events interleaved with component lifecycle hooks. This is why we wouldn't have caught it as there's no documentation or tests that cover this sequence.

That being said, it sounds like there are use cases you need to solve. I agree just subscribing in the `constructor` is not ideal and it would be correct to use `ngOnInit` and `ngOnDestroy`. So, that being said, what is the use case? It appears you were getting the `NavigationEnd` event on the sub-component level. What were you assuming this meant in terms of application state? Maybe there's another event that should be published to support a use case if it's a common one.

**Comment by dahaupt** [comment]
@jasonaden We are using it to read data of an active route when changing the routes (Event **NavigationEnd**) for a "MenuComponent". It's not working anymore after initial startup. When navigating again it's working as expected.

` {path: 'customer/account', component: AccountComponent, data: {name: 'Custom Title'}}`

**MenuComponent**:

```typescript
ngOnInit() {
    this.router.events.filter(event => event instanceof NavigationEnd)
      .subscribe(
        event => {
          let currentRoute = this.route.root; // route is an instance of ActiveRoute
          let data;
          while (currentRoute.children[0] !== undefined) {
            currentRoute = currentRoute.children[0];
          }
          data = currentRoute.snapshot.data;
          this.currentPage = data.name;
          if (data.name) {
            this.navigationService.setTitle(data.name);
          } else {
            this.navigationService.setDefaultTitle();
          }
        }
      );
  }
```

**Comment by kemsky** [comment]
It breaks things even if you subscribe in constructor. Once again, there is now way to change timing of events without breaking existing code.

It should be documented right now in change log as breaking change or reverted and postponed until 5.0.

**Comment by francois-appliware** [comment]
I too have the same problem with how router fires event.
I actually have 2 problems that are linked to the router event:

- I have a load indicator for lazy module when navigating to a page not loaded yet. (The kind of loader Youtube shows when navigating from videos to videos). I need to get the NavigationEnd to hide it, but I can't anymore as I subscribe in the ngOnInit(). (which is kind of how Angular wants us to do)

- I use a lot rxjs and some of my Observables just won't work anymore. Operators like 'combineLatest', waiting for the event to fire, too late to catch it.

Is there any way to have some kind of replay subject or anything like that to get the events onInit ?
Parts of my app are completely blocked because of this :(

**Comment by jasonaden** [comment]
Thanks for the feedback on this issue. We're discussing this and would appreciate some more use cases. What we're looking for would be a real-world use case. Could one of you (or multiples) create another plunker that demonstrates your problem in a more real-world scenario?

Thanks!

**Comment by vicb** [comment]
It's used to be that all component would be change detected during the navigation (1) and once again after the navigation (2).

During (1) the `ngOnInit` hook would be called and that's before the `NavigationEnd` event is triggered.

In the latest version `ngOnInit` is only called during (2) as we do not run extra change detection (1). Because (2) happens after the event is triggered you would miss it for the navigation that created the component - but not for subsequent navigations.

> It breaks things even if you subscribe in constructor

It should not. `ngOnInit` is called after the navigation but still components are created during the navigation (provided they are stamped out, ie oustide of a `ngIf="false"`). Do you have a plunker ?

also see

> Placing the call in the constructor of the sub component fixes my issue, but I am curious as to why this was changed. I would like to use ngOnInit and ngOnDestroy for most initialization and destruction procedures.

A question I would is have is when subscribing to `NavigationEnd` in a `ngOnInit` is a useful pattern ? (We are looking for use cases here).
From the previous comments, the component subscribing to the event assumes that `ngOnInit` will be called during the navigation that creates its router outlet. What if the component is inside an `ngIf=false` ? In that case `ngOnInit` will only be called when the condition becomes true (after the event has fired).

@dahaupt you probably can do something like

```ts
ngOnInit() {
   this.updateTitle(this.route.root);

    this.router.events.filter(event => event instanceof NavigationEnd)
      .subscribe(e => this.updateTitle(this.route.root));
  }

updateTitle(root) {
  // ...
}
```

@francois-appliware you could probably use something along the lines of the previous ex. One question though is why would it be the responsibility of your child component to hide the loading indicator ?

Once again if you have concrete use case/repro we can take a look at them. Also we'll document this better.

**Comment by fillmorejl** [comment]
In my case I am not even listening for route events, I am just calling subject.next(this) (rxjs) from ngOnInit to pass an instance of the child to the parent so the parent can read the child's title and it's actions collection to build buttons in the master header, the parent does not get updated any longer.  When I click on the routerLink again, even though navigation does not happen, the parent then updates accordingly.  I will try and provide a Plunker soon, busy weekend.

**Comment by francois-appliware** [comment]
Hello again,

**TL;DR: I don't use Router.events anymore, or I put the sub in the constructor and that's fine. (For my use case !!)**

1. I use some kind of MasterPage component that holds the menu of my app. I have 2 menus, one for user, one for admin. Some items of a menu come from lazy loaded modules. When navigating to a menu item that references a lazy module, I can show the load indicator to signal my user the navigation is in progress. I swithed the subscription in the constructor and did as follow:
```typescript
constructor(private state: ApplicationState,
        private loginService: LoginService, private router: Router) {
        this.routerSub = this.router.events.subscribe(event => this.navigationInterceptor(event));
    }

    ngOnInit(): void {
        if (this.routerSub == null)
            this.routerSub = this.router.events.subscribe(event => this.navigationInterceptor(event));
    }

    ngOnDestroy(): void {
        if (this.routerSub != null) {
            this.routerSub.unsubscribe();
            this.routerSub = null;
        }
    }
```
The navigation interceptor being:
```typescript
private navigationInterceptor(event: RouterEvent): void {
        if (event instanceof NavigationStart)
            this.state.startGlobalWork();
        if (event instanceof NavigationEnd || event instanceof NavigationError || event instanceof NavigationCancel)
            this.state.stopGlobalWork();
    }
```
My state holds observables that are used with the async pipe. My global indicator is tested with a ngIf. The start/stop GlobalWork pushes either true or false in the underlying subject, used as an observable in my view.

My master page is used in 2 components that have nothing expect the master page and the menu links.
My master page gets destroyed and recreate when I swith from Home / Admin. With the above code it seems to work fine.


2. I used to use the router.events for component navigation but I don't anymore. Instead, I use a custom NavigationService (which is just a Subject with getter as Observable) and call it inside the ngOnInit() on my child component. That way I don't rely on Router.events anymore.

**Comment by kemsky** [comment]
I have the following structure:

* App component with router outlet.
* Main component with router outlet (created in outlet of app component).
* Menu component with `*ngFor` which renders several menu items (created in outlet of main component).
* Each menu item (a directive) subscribes to router events in constructor.

App component -> router outlet -> main component -> router outlet -> menu component 

Application is opened first time. Before 4.2 router, directive constructor was  called before navigation end event, since 4.2 directive constructor is called after navigation end event, so event is missed.

Before 4.2:
* Create App
* Navigation Start
* Route Recognized
* Create Main
* Create Menu
* Create Menu item directives
* Navigation End

Since 4.2:
* Create App
* Navigation Start
* Route Recognized
* Create Main
* Create Menu
* Navigation End
* Create Menu item directives

**Comment by jasonaden** [comment]
@fillmorejl Thanks for the additional info. Yes, a Plunker showing the problem would really help in this case.

@francois-appliware Thanks for the code samples. It sounds like you moved your global loading indicator (or data your indicator can bind to) to the global/application level. This sounds like a better implementation.

Your note said you no longer use the router events, which means as far as you're concerned you are no longer affected by this issue, correct?

@kemsky Thanks for the explanation of what your implementation looks like and the sequence of events. Would you be able to provide a Plunker that demonstrates what you're doing? Short of that, could you explain in more detail why the Menu components are subscribing to router events? What are they using those events for?

**Comment by kemsky** [comment]
Adding `*ngFor` to the plunker in the first post immediately shows the problem:
https://plnkr.co/edit/0K6DZ319YaHZialULOUm?p=preview

**Comment by francois-appliware** [comment]
@jasonaden I **do** use the router.events in my master page component which subscribes to it in its constructor, with checks in the ngOnInit() and ngOnDestroy() to be sure that 
  1. I am subscribed to it, 
  2. I unsubsribe to it when the component gets destroyed.

That's the explanation number 1 of my previous post. Doing so, with my indicator being app wide, it does seem to work as expected. (If not I will make a plunker, but _for now it's okay_)


My explanation number 2 (end of the previous post) does not use the router anymore and instead, use a custom service I created (basically a Subject from rxjs). This service is used to fire an event when a child gets ngOnInit(), for the parent to subscribe to it and do something with it.
I did not have to use such a service before because I could rely on the router.events subscription.


**So yes, I am not concerned by this issue anymore, but as I still use it, I am very interested in seeing how things evolve !**

Cheers!

**Comment by Da13Harris** [comment]
This has certainly been a breaking change in our enterprise application.

In our case, we have tabs that can be independently navigated. We track the current route of each tab in an array in a global service. We use the NavigationEnd event to store the route of each tab as the user navigates within said tab.

For the time being, I was able to fix our issue by moving the event subscription into the constructor.

I changed this:
```
    constructor(
        private router: Router,
        private route: ActivatedRoute,
        public tabsService: TabsService
    ) { }

    ngOnInit() {
        // Set current tab based on route, and reflect it in the tab service
        this.currentTab = Number(this.route.snapshot.params['id']);
        this.tabsService.setCurrentTab(this.currentTab);
        // Each time the router completes navigation, set the current tab's route
        this.router.events
            .takeUntil(this.ngUnsubscribe)
            .subscribe(
            event => {
                if (event instanceof NavigationEnd) {
                    this.tabsService.tabs[this.tabsService.currentTabIndex].route = this.router.url;
                }
            }
            );
    }
```


to this:
```
    constructor(
        private router: Router,
        private route: ActivatedRoute,
        public tabsService: TabsService
    ) {
        // Must detect routes in constructor, because they no longer occur after ngOnInit
        this.storeRoutesAfterNavigation();
    }

    storeRoutesAfterNavigation() {
        // Each time the router completes navigation, set the current tab's route
        this.router.events
            .takeUntil(this.ngUnsubscribe)
            .subscribe(
                event => {
                    if (event instanceof NavigationEnd) {
                        this.tabsService.tabs[this.tabsService.currentTabIndex].route = this.router.url;
                    }
                }
            );
    }

    ngOnInit() {
        // Set current tab based on route, and reflect it in the tab service
        this.currentTab = Number(this.route.snapshot.params['id']);
        this.tabsService.setCurrentTab(this.currentTab);
    }
```

**Comment by cortopy** [comment]
I've also been bitten by what is clearly a breaking change. In my case I needed to trigger a scroll to an element when user had selected to navigate to it from another page.

Before (simplified):
```javascript
ngAfterViewInit() {
    this.router.events
      .filter(ev => ev instanceof NavigationEnd)
      .subscribe(ev => {
        const queryParams = this.route.snapshot.queryParams;
        if (queryParams["myElementID"]) {
          // do the scroll
        }
     );
  }
```
And the extremely HACKISH version after:

```javascript
private viewReady: BehaviorSubject<boolean> = new BehaviorSubject(false);

constructor(private router: Router, private route: ActivatedRoute) {
   this.router.events
      .filter(ev => ev instanceof NavigationEnd)
      .switchMapTo(this.viewReady)
      .filter(ready => !!ready)
      .subscribe(ready => {
        const queryParams = this.route.snapshot.queryParams;
        if (queryParams["myElementID"]) {
          // do the scroll
        }
     );
}

ngAfterViewInit() {
    this.viewReady.next(true);
  }
```

This is so hackish that more than a breaking change, this looks to me like a regression. Or is there anything I'm missing?

**Comment by bleuscyther** [comment]
the best solution so far is @Da13Harris , well its based on many comments on top.
Something to note:

@Input property isn't initialized until view  ready so  some operations will need to happen on ngOnInit().

the constructor may store the value that will be use durring  ngOnInit ( in a variable/attribute for example)

**Comment by argasek** [comment]
It looks like our app has been affected by this issue as well. I our scenario, we use the ngOnInit subscription to router events to dynamically create a secondary navigation depending on which component (and, in consequence, route) is active.

Basically, each feature component (which is bound to particular route) has a respective resolver. This resolver puts the secondary navigation configuration (think of an array of anchors) in the route's `data`.

The feature components are nested in a main container component. In the `ngOnInit()` of the main container, we subscribe to the route navigation events, filter out NavigationEnd events and look if the active route has any secondary navigation configuration in the `data`. If such configuration is found, we emit it as an event to the navigation configuration stream.

Another component nested in the main container is Secondary Navigation component itself, which (also in `ngOnInit()`) subscribes to the previously mentioned configuration stream and - if event was caught - updates the set of anchors in the DOM.

(We have some more complex scenarios, eg. components not having the respective resolvers, but instead setting the navigation configuration based on async response received from server, etc.) 

So, the order of events happening is like that (taken directly from app's console.log):
```
Feature Component: Resolving (setting) Secondary Navigation configuration
Main Container ngOnInit: subscribe to router events in order to filter them out and cast updated navigation configuration (navigationConfigurationSubscription)
Secondary Navigation ngOnInit: subscribe to Secondary Navigation configuration changes
navigationConfigurationSubscription: navigation ended, casting extracted route data to listeners
Secondary Navigation listener: received updated navbar configuration
```
After an update from Angular 4.1.3 to 4.3.6, I get only these in the console:
```
Feature Component: Resolving (setting) Secondary Navigation configuration
Main Container ngOnInit: subscribe to router events in order to filter them out and cast updated navigation configuration (navigationConfigurationSubscription)
Secondary Navigation ngOnInit: subscribe to Secondary Navigation configuration changes
```
Moving the router events and configuration changes subscriptions to the respective constructors workarounds this issue, but from my perspective such breaking changes should not take place in the 4.x line - especially it wasn't announced anywhere clearly. (Plus, I handle all the subscribe/unsubscribe elegantly in the ngOnInit / ngOnDestroy, now I have some code in the constructor and some in ngOnDestroy, which is weird).

**Comment by sgwatgit** [comment]
Solutions that come to mind

1. Subscribe to the `Router.events` in another wrapping service that uses a `BehaviorSubject` to store the last event. Deeply nested child components could subscribe the wrapping service's `events` `Observable` to get the last event that happened if it happened before their `ngOnInit` was called.

2. Depending on what you need out of the event, use `.startWith` and fill it with data from the `Router.routerState.snapshot` or dummy data if you only need the router event to kick off some other stream. This seems like it might work for many since most users want to be alerted when they reach `NavigationEnd`.

**Comment by klinki** [comment]
Our application has been also hit by this issue, which means we have to move a lot of code from `ngOnInit` to `constructor`. This is certainly very breaking change, since our app doesn't work with this change. 

I'm really unhappy about this situation (yet another reminder of many issues I had with Angular Router. And from my experience, I can probably expect many more...).

**Comment by JohnGalt1717** [comment]
Also here. If you're using prerender.io instead of SSR (which currently has a separate bug in http requests so we've just given up) you need to set window.prerenderReady = true. The perfect spot for this should be the NavigationEnd which should fire after the ngAfterViewInit of the component that was loaded by the route change.

However in Angular 5.x this fires almost immediately when the component isn't loaded thus giving a false positive to the indexer for prerender.io.

**Comment by Wingnod** [comment]
I encountered this issue recently, and decided to wrap the router.events in a ReplaySubject.

private _routerEvents = new ReplaySubject<NavigationEnd>();

constructor(private router: Router) {
    this.router.events.subscribe(this._routerEvents);
}
This should effectively make the subject act as a hot observable, receiving emissions from the events observable. Since it is a replay subject, each time it emits it should have a history of all previous emissions.

this._routerEvents
      .pipe(filter(e => e instanceof NavigationEnd))
      .subscribe((e) => {
            console.log(e);
      });
The resulting values on first load of the component (that subscribes to the router events subject) should have all the events you would expect, and of course, continue to emit events thereafter as a kind of proxy for the events observable on router.

I consider this cleaner than doing everything in the constructor at least.

**Comment by ftsai-castlight** [comment]
Though I understand that this isn't a breaking change it would be great to create some API to address this issue.

We have been extensively using route data to drive page specific changes, (for example, page titles or page specific changes in the header and footer). So components subscribe to router.events and on navigation change, check the data and set state. We've been seeing bugs and to work around this issue we also wrapped the routers.events in a replaySubject like Wingnod did and replaced all of the router.events subscriptions with subscriptions to the subject. 

It would be great to have perhaps a different property, something like router.navigationEndEvents which is backed up by a replay subject to eliminate this issue. Otherwise, anyone depending on the use of route data has to create a replaySubject to ensure that the routing data is getting updated.

**Comment by det-jscott116** [comment]
Hi,

I have had a brief read through this issue, but I'm not sure that my issue is exactly the same. 

I recently looked at cleaning up some of my code and moved an angular `Router` subscription event from the `constructor` of a component into the `ngOnInit()` function.

As a result, the `NavigationEnd` event is no longer picked up. After moving it back into the `constructor` its back to normal.

There is nothing fancy happening in my code - its a really simple implementation;

```
  constructor() {
  }

  ngOnInit() {
    this.watchRoute();
  }

  private watchRoute() {
    this.router.events
      .pipe(
        takeUntil(this.ngUnsubscribe)
      )
      .subscribe((event) => {
        this.addButtonEnabled = false;
        this.addAnnotationButtonEnabled = false;

        console.log(event);

        // The following block is never executed as "NavigationEnd" events are not received
        if (event instanceof NavigationEnd) {
          if (event.url.startsWith('/' + AppRoutes.GAP)) {
            this.addButtonEnabled = true;

            if (new RegExp('/' + AppRoutes.GAP + '\\/\\d+').test(event.url)) {
              this.addAnnotationButtonEnabled = true;
            }
          }
        }
      });
  }
```

As mentioned after moving `this.watchRoute();` into the `constructor` everything is working as expected. 

I have no issue calling this function from the constructor, but I recently had been reading about best-practice for Angular and found multiple sources recommending that no code should be run in the constructor - other than for basic things such as initialising class properties, etc. My feeling is that this change should not affect the result of a subscription - how is this even possible??

**Comment by vinagreti** [comment]
For those who are still facing this problem, I suggest you to use `shareReplay()` to replay the last event outside the constructor;

Lets say we are accessing the `/news` route path;

```javascript
  constructor(
    public router: Router,
  ) {
    this.routePath$ = this.router.events.pipe(
      filter(event => event instanceof NavigationEnd),
      map((event: any) => event.url),
      distinctUntilChanged(),
      // replay sthe latest NavigationEnd emissions on subscription: https://www.learnrxjs.io/operators/multicasting/sharereplay.html
      shareReplay(),
    );

    // force shareReplay to be able to listen to the NavigationEnd event that happens only in constructor
    this.routePath$.pipe(take(1)).subscribe();
  }

  ngOnInit() {
    // logged: '/news'
    this.routePath$.subscribe(console.log);
  }
```

This is very usefull when using `changeDetection: ChangeDetectionStrategy.OnPush` and you want to use observables to trigger change detection.

This way we persist the emissions that happened before the constructor's end.

**Comment by atscott** [comment]
Hi all, there's a lot to unpack here and I'm not entirely sure how to triage this issue at the moment. There seems to be some workaround available, and as Jason mentioned 
> there isn't a guarantee of the sequence of router events interleaved with component lifecycle hooks. This is why we wouldn't have caught it as there's no documentation or tests that cover this sequence
(https://github.com/angular/angular/issues/17473#issuecomment-308848113)

Victor also responded to several points here
https://github.com/angular/angular/issues/17473#issuecomment-309168922
and there seem to be several workarounds, including `shareReplay` or `ReplaySubject`, adding the subscription to the constructor rather than `ngOnInit`, or adding the subscription to the root component. In addition, it seems like several of these issues could be resolved by using the `ActivatedRoute` and subscribing to the Observables provided there. 

Given that a lot of the issues described in this bug are several years old and that I don't know what API endpoints were available at the time, I can't tell what is still relevant. I'll keep this open for now, as this is potentially an issue that needs to be addressed with better documentation to cover use-cases that don't currently have good recommendations.

**Comment by atscott** [comment]
Closing as obsolete.
